talk/expert/cpp20concepts.tex

```diff
@@ -0,0 +1,199 @@
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Requirements and concepts}
+    \begin{block}{Concepts}
+      \begin{itemize}
+        \item When a given set of requirements may be reused often, one can gather them in a concept.
+      \end{itemize}
+    \end{block}
+    \begin{exampleblock}{With {\it concept}}
+      \scriptsize
+      \begin{cppcode*}{}
+      template< typename T>
+      concept MyFloatingPoint =
+        (std::is_floating_point_v<T>) &&
+        (std::numeric_limits<T>::epsilon()>0) ;
+
+      template<typename T>
+      requires MyFloatingPoint<T>
+      bool equal( T e1, T e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Requirements and concepts}
+    \begin{block}{Within the template head}
+      \begin{itemize}
+        \item The concept can also be directly used instead of {\it typename} within the template parameter list.
+      \end{itemize}
+    \end{block}
+    \begin{exampleblock}{Instead of {\it typename}}
+      \scriptsize
+      \begin{cppcode*}{}
+      template<MyFloatingPoint T>
+      bool equal( T e1, T e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Requirements and concepts}
+    \begin{block}{With abbreviated function templates}
+      \begin{itemize}
+        \item The concepts can be used together with {\it auto} in the abbreviated function templates.
+      \end{itemize}
+    \end{block}
+    \begin{exampleblock}{With {\it auto}}
+      \scriptsize
+      \begin{cppcode*}{}
+      bool equal( MyFloatingPoint auto e1, MyFloatingPoint auto e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Requirements and concepts}
+    \begin{block}{Standard concepts}
+      \begin{itemize}
+        \item Writing a bug-proof concept is an expert task.
+        \item Prefer the ones provided by the standard library.
+        \item Not surprisingly, there is one for floating point numbers.
+      \end{itemize}
+    \end{block}
+    \begin{exampleblock}{With {\it std::}}
+      \scriptsize
+      \begin{cppcode*}{}
+      #include <concepts>
+
+      bool equal( std::floating_point auto e1, std::floating_point auto e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Requirements and concepts}
+    \begin{block}{Within a function body}
+      \begin{itemize}
+        \item Concepts are also usable wherever a boolean is expected.
+        \item Since they are evaluated at compile-time, they can appear in a {\it if constexpr} condition.
```

Comment:
What about e.g. `\texttt` or `\mintinline`? Like this, the keywords look a bit more like code.

Reply:
@hageboeck \mintinline did not work. \texttt is not very distinguishable from the rest of the text. I failed also with things like \texttt{\textbf{taratata}}... but I am not at ease with latex and loosing much time waiting the latex compilation after each try. Thinking of going back to \it...

Reply:
@sponce By th eway, I think the chosen default policy, "sans-serif" is ok for the slides titles, but some "serif" policy would be better for the text. I have no idea how to change that in latex ;) The general typographic rule I have in mind is that  one should choose sans-serif policy for the titles, so that it "stop the look", and that a serif policy should be used for usual text that one should read smoothly. As a side effect, the code keywords that we would like to put in \texttt would be more visibles.